From 2d787de9a05201d18c737dbf5b1c16c42da831f0 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 23 Jun 2020 10:13:08 +0300
Subject: [PATCH] [SW-14252] ppboot: goya2: initialize arc0

This patch adds support to perform required initialization for
ARC0 before copying zephyr images in ICCM and DCCM.

As part of initialization, reset vector addr is set to 0x0 location of
ICCM and ARC is brought out of reset. Note that ARC remains in halt
state on out of reset.

Change-Id: I83a850055d9aab50fb7ba2009ebaec51995478e6
---

diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index 68dc2c9..1dda76b 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -8,6 +8,7 @@
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio1_regs.h>
 #include <psoc_qspi_regs.h>
+#include <psoc_arc0_cfg_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_wrap_regs.h>
 #include <pcie_core_regs.h>
@@ -251,10 +252,24 @@
 	regmap_writel(0x0, mmPSOC_GLOBAL_CONF_BOOT_SEQ_EXT_LD);
 }
 
+static void init_arc0(void)
+{
+	/*
+	 * Program Reset vector address for ARC0 to ICCM 0x0 location
+	 * where vector table first entry will be available and then
+	 * bring ARC0 out of reset. ARC when out of reset is by default
+	 * in halt state.
+	 */
+	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
+	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
+}
+
 void hl_plat_init(void)
 {
 	/* Functionality used to be done in BTL */
 	init_prstn();
+
+	init_arc0();
 }
 
 void spi_ssi_cs0_get(unsigned int *mask, unsigned int *data_offset,
diff --git a/goya2_arc/common.h b/goya2_arc/common.h
index 1781fc0..1805014 100644
--- a/goya2_arc/common.h
+++ b/goya2_arc/common.h
@@ -27,6 +27,9 @@
 #define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI	BIT(6)
 #define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
 
+#define ARC_RST_VALUE		0x110
+#define ARC_RST_VEC_ADDR	0x0
+
 /*
  * extracts "num" number of bits from position "pos" in the
  * given number "value"
